File: src/backend/forensics/audio/spectrogram_decimate.py
"""Decimacao max-pool do espectrograma para exibicao Plotly (performance)."""

from __future__ import annotations

import numpy as np

DEFAULT_MAX_TIME_BINS = 2000
DEFAULT_MAX_FREQ_BINS = 512


def decimate_spectrogram_max_pool(
    magnitude_db: np.ndarray,
    times: np.ndarray,
    freqs: np.ndarray,
    *,
    max_time_bins: int = DEFAULT_MAX_TIME_BINS,
    max_freq_bins: int = DEFAULT_MAX_FREQ_BINS,
) -> tuple[np.ndarray, np.ndarray, np.ndarray, dict]:
    """
    Reduz matriz tempo-frequencia para o grafico sem alterar o STFT completo.

    Usa max-pool por bloco para preservar picos de energia (tons, transientes).
    """
    z = np.asarray(magnitude_db, dtype=np.float64)
    times = np.asarray(times, dtype=np.float64)
    freqs = np.asarray(freqs, dtype=np.float64)

    n_rows, n_cols = z.shape
    meta = {
        "decimated": False,
        "full_shape": [int(n_rows), int(n_cols)],
        "display_shape": [int(n_rows), int(n_cols)],
        "row_pool_factor": 1,
        "col_pool_factor": 1,
        "max_time_bins": max_time_bins,
        "max_freq_bins": max_freq_bins,
    }

    if n_cols <= max_time_bins and n_rows <= max_freq_bins:
        return z, times, freqs, meta

    col_factor = max(1, int(np.ceil(n_cols / max_time_bins)))
    row_factor = max(1, int(np.ceil(n_rows / max_freq_bins)))
    n_cols_out = int(np.ceil(n_cols / col_factor))
    n_rows_out = int(np.ceil(n_rows / row_factor))

    pad_rows = n_rows_out * row_factor - n_rows
    pad_cols = n_cols_out * col_factor - n_cols
    pad_value = float(np.nanmin(z)) if z.size else -120.0
    zp = np.pad(z, ((0, pad_rows), (0, pad_cols)), mode="constant", constant_values=pad_value)
    z_out = zp.reshape(n_rows_out, row_factor, n_cols_out, col_factor).max(axis=(1, 3))

    col_centers = np.minimum(np.arange(n_cols_out) * col_factor + col_factor // 2, n_cols - 1)
    row_centers = np.minimum(np.arange(n_rows_out) * row_factor + row_factor // 2, n_rows - 1)
    times_out = times[col_centers]
    freqs_out = freqs[row_centers]

    meta.update(
        {
            "decimated": True,
            "display_shape": [int(n_rows_out), int(n_cols_out)],
            "row_pool_factor": row_factor,
            "col_pool_factor": col_factor,
        }
    )
    return z_out, times_out, freqs_out, meta

File: src/frontend/src/lib/spectrogramDecimate.ts
/** Max-pool decimation (port of spectrogram_decimate.py) for client-side display. */

export const DEFAULT_MAX_TIME_BINS = 2000;
export const DEFAULT_MAX_FREQ_BINS = 512;

export interface SpectrogramDecimationMeta {
  decimated: boolean;
  full_shape: [number, number];
  display_shape: [number, number];
  row_pool_factor: number;
  col_pool_factor: number;
}

function min2d(rows: number[][]): number {
  let m = Infinity;
  for (const row of rows) {
    for (const v of row) {
      if (v < m) m = v;
    }
  }
  return Number.isFinite(m) ? m : -120;
}

export function decimateSpectrogramMaxPool(
  magnitudeDb: number[][],
  times: number[],
  frequencies: number[],
  maxTimeBins = DEFAULT_MAX_TIME_BINS,
  maxFreqBins = DEFAULT_MAX_FREQ_BINS
): {
  magnitude_db: number[][];
  times: number[];
  frequencies: number[];
  meta: SpectrogramDecimationMeta;
} {
  const nRows = magnitudeDb.length;
  const nCols = nRows > 0 ? magnitudeDb[0].length : 0;
  const meta: SpectrogramDecimationMeta = {
    decimated: false,
    full_shape: [nRows, nCols],
    display_shape: [nRows, nCols],
    row_pool_factor: 1,
    col_pool_factor: 1,
  };

  if (nCols <= maxTimeBins && nRows <= maxFreqBins) {
    return { magnitude_db: magnitudeDb, times, frequencies, meta };
  }

  const colFactor = Math.max(1, Math.ceil(nCols / maxTimeBins));
  const rowFactor = Math.max(1, Math.ceil(nRows / maxFreqBins));
  const nColsOut = Math.ceil(nCols / colFactor);
  const nRowsOut = Math.ceil(nRows / rowFactor);
  const padValue = min2d(magnitudeDb);

  const zOut: number[][] = [];
  for (let ri = 0; ri < nRowsOut; ri++) {
    const row: number[] = [];
    for (let ci = 0; ci < nColsOut; ci++) {
      let peak = padValue;
      for (let dr = 0; dr < rowFactor; dr++) {
        const r = ri * rowFactor + dr;
        if (r >= nRows) continue;
        for (let dc = 0; dc < colFactor; dc++) {
          const c = ci * colFactor + dc;
          if (c >= nCols) continue;
          peak = Math.max(peak, magnitudeDb[r][c]);
        }
      }
      row.push(peak);
    }
    zOut.push(row);
  }

  const timesOut: number[] = [];
  for (let ci = 0; ci < nColsOut; ci++) {
    const idx = Math.min(ci * colFactor + Math.floor(colFactor / 2), nCols - 1);
    timesOut.push(times[idx]);
  }

  const freqsOut: number[] = [];
  for (let ri = 0; ri < nRowsOut; ri++) {
    const idx = Math.min(ri * rowFactor + Math.floor(rowFactor / 2), nRows - 1);
    freqsOut.push(frequencies[idx]);
  }

  meta.decimated = true;
  meta.display_shape = [nRowsOut, nColsOut];
  meta.row_pool_factor = rowFactor;
  meta.col_pool_factor = colFactor;

  return { magnitude_db: zOut, times: timesOut, frequencies: freqsOut, meta };
}

export function applySpectrogramDisplayOptions(
  full: {
    times: number[];
    frequencies: number[];
    magnitude_db: number[][];
  },
  decimate: boolean
) {
  if (!decimate) {
    const nRows = full.magnitude_db.length;
    const nCols = nRows > 0 ? full.magnitude_db[0].length : 0;
    return {
      ...full,
      meta: {
        decimated: false,
        full_shape: [nRows, nCols] as [number, number],
        display_shape: [nRows, nCols] as [number, number],
        row_pool_factor: 1,
        col_pool_factor: 1,
      },
    };
  }
  const out = decimateSpectrogramMaxPool(full.magnitude_db, full.times, full.frequencies);
  return { ...out, meta: out.meta };
}
